wgsl/index.bs

```diff
@@ -1757,67 +1763,58 @@ following table:
+      <td>|N| &times; [=roundUp=]([=AlignOf=](|E|), [=SizeOf=](|E|))
   <tr><td>array<|E|><br>
       <td>[=AlignOf=](|E|)
-      <td>N<sub>runtime</sub> * [=roundUp=]([=AlignOf=](|E|), [=SizeOf=](|E|))<br><br>
-          Where N<sub>runtime</sub> is the runtime-determined number of elements of |T|
+      <td>N<sub>runtime</sub> &times; [=roundUp=]([=AlignOf=](|E|),[=SizeOf=](|E|))<br><br>
+          where N<sub>runtime</sub> is the runtime-determined number of elements of |T|
 </table>
 
 
-#### Structure Layout Rules ####  {#structure-layout-rules}
+#### Structure Member Layout ####  {#structure-member-layout}
 
-Issue: [#2493](https://github.com/gpuweb/gpuweb/issues/2493) Portions of this section are redundant with other sections.
+Each structure |S| member M<sub>i</sub> has a size and alignment, denoted
+by [=SizeOfMember=](|S|, M<sub>i</sub>) and [=AlignOfMember=](|S|, M<sub>i</sub>), respectively.
+The member sizes and alignments are used to calculate each member's byte offset from the start of the structure,
+as described in [[#internal-value-layout]].
 
-Each structure |S| member M<sub>N</sub> has a size and alignment value, denoted
-by [=SizeOfMember=](|S|, M<sub>N</sub>) and [=AlignOfMember=](|S|, M<sub>N</sub>), respectively.
-The member sizes and alignments are used to calculate each member's byte offset from the start of the structure.
-
-Structures member size and alignment values default to the member type `T`'s
-[=SizeOf=](T) and [=AlignOf=](T) values.
-
-If a structure member is decorated with the [=attribute/size=] decoration, then the structure
-member will use the value of the decoration for its size instead of its type's size.
-
-If a structure member is decorated with the [=attribute/align=] decoration, then the structure
-member will use the value of the decoration for its alignment instead of its type's alignment.
-
-The first structure member always has a zero byte offset from the start of the
-structure.
+<p algorithm="structure member size">
+  [=SizeOfMember=](|S|, M<sub>i</sub>) is |k| if member M<sub>i</sub> has attribute [=attribute/size=](|k|).
+  Otherwise, it is [=SizeOf=](|T|) where |T| is the type of member M<sub>i</sub>.
+</p>
 
-Subsequent members have the following byte offset from the start of the structure:
-<p algorithm="structure member offset">
-  [=OffsetOfMember=](|S|, M<sub>N</sub>) = [=roundUp=]([=AlignOfMember=](|S|, M<sub>N</sub>), [=OffsetOfMember=](|S|, M<sub>N-1</sub>) + [=SizeOfMember=](|S|, M<sub>N-1</sub>)<br>
-  Where M<sub>N</sub> is the current member and M<sub>N-1</sub> is the previous member
+<p algorithm="structure member alignment">
+  [=AlignOfMember=](|S|, M<sub>i</sub>) is |k| if member M<sub>i</sub> has attribute [=attribute/align=](|k|).
+  Otherwise, it is [=AlignOf=](|T|) where |T| is the type of member M<sub>i</sub>.
 </p>
 
-Structure members must not overlap. If a structure member is decorated with the
+If a structure member is decorated with the
```

Comment:
I don't see a similar restriction on `align`. Do we intend to allow packed structures? This seems to be the existing state of affairs.

I would consider either moving or repeating this constraint in the definition of the size attribute (and align if a similar restriction is added).

Reply:
> Do we intend to allow packed structures? This seems to be the existing state of affairs.

Good question.  It's a bit roundabout, but RequiredAlignOf ends up enforcing that actual alignments are at least as stringent as the defaults.


> Structure members of type T must have a byte offset from the start of the structure that is a multiple of the RequiredAlignOf(T, C) for the storage class C:

> OffsetOfMember(S, M) = k × RequiredAlignOf(T, C)

Maybe this is worth a note.

Of course, I could see an extension later that relaxes that.

Reply:
> I would consider either moving or repeating this constraint in the definition of the size attribute 

Makes a lot of sense.